tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/AUC.java

```diff
@@ -0,0 +1,1075 @@
+   *
+   * @param labels shape (N, Cx, L1?) where N is the number of examples, Cx is zero or more class
+   *     dimensions, and L1 is a potential extra dimension of size 1 that would be squeezed. Will be
+   *     cast to {@code <T>}. If {@link #multiLabel} or if {@link #labelWeights} {@code != null
+   *     }, then Cx must be a single dimension.
+   * @param predictions the predictions shape (N, Cx, P1?). Will be cast to {@code T}.
+   * @param sampleWeights sample weights to be applied to values, may be null. Will be cast to
+   *     {@code <T>}.
+   * @return a List of Operations to update the metric state
+   */
+  @Override
+  @SuppressWarnings("unchecked")
+  public List<Op> updateStateList(
+      Operand<? extends TNumber> labels,
+      Operand<? extends TNumber> predictions,
+      Operand<? extends TNumber> sampleWeights) {
+    Ops tf = getTF();
+    Operand<T> tLabels = cast(tf, labels, type);
+    Operand<T> tPredictions = cast(tf, predictions, type);
+    Operand<T> tSampleWeights = sampleWeights != null ? cast(tf, sampleWeights, type) : null;
+    List<Op> updateOperations = new ArrayList<>();
+    Map<ConfusionMatrixEnum, Assign<T>> varInitializers = Collections.EMPTY_MAP;
+    if (!initialized) {
+      varInitializers = build(tPredictions.shape());
+    }
+    if (isMultiLabel() || getLabelWeights() != null) {
+      // labels should have shape (number of examples, number of labels).
+      List<SymbolicShape<? extends TNumber>> symbols = new ArrayList<>();
+      symbols.add(new SymbolicShape<>(tLabels, "N", "L"));
+      if (isMultiLabel()) {
+        // TP, TN, FP, and FN should all have shape
+        // (number of thresholds, number of labels).
+        symbols.add(new SymbolicShape<>(truePositives, "T", "L"));
+        symbols.add(new SymbolicShape<>(falsePositives, "T", "L"));
+        symbols.add(new SymbolicShape<>(trueNegatives, "T", "L"));
+        symbols.add(new SymbolicShape<>(falseNegatives, "T", "L"));
+      }
+      if (getLabelWeights() != null) {
+        symbols.add(new SymbolicShape<>(getLabelWeights(), "L"));
+      }
+      updateOperations.addAll(
+          MetricsHelper.assertShapes(tf, symbols, "Number of labels is not consistent."));
+    }
+
+    Map<ConfusionMatrixEnum, Variable<T>> confusionMatrix = new HashMap<>();
+    confusionMatrix.put(ConfusionMatrixEnum.TRUE_POSITIVES, truePositives);
+    confusionMatrix.put(ConfusionMatrixEnum.FALSE_POSITIVES, falsePositives);
+    confusionMatrix.put(ConfusionMatrixEnum.TRUE_NEGATIVES, trueNegatives);
+    confusionMatrix.put(ConfusionMatrixEnum.FALSE_NEGATIVES, falseNegatives);
+
+    // Only forward labelWeights to update_confusion_matrix_variables when
+    // multiLabel is false. Otherwise the averaging of individual label AUCs is
+    // handled in AUC.result
+    updateOperations.addAll(
+        MetricsHelper.updateConfusionMatrixVariables(
+            tf,
+            confusionMatrix,
+            varInitializers,
```

Comment:
The initializers only get attached to the confusion variables  when the AUC  `updateStates` is first run. After that the CM variables are not not reinitialized until `resetStates` is called. This is done this way because the CM variables cannot be created until `build` is called with the input shapes on the first invocation of `updateStates`. This is controlled by the `boolean initialized `member.